### PR TITLE
first version of persistent lock utitlity and intergration

### PR DESCRIPTION
**Pull request by azhao155**
first version of persistent lock utility for modsec
In Nginx, modsecurity_init will be called whenever there is config update and never released, so it will cause semaphore leak, and apr run time library doesn't have persistent lock api, so we implement this utility. 

Have tested on the multiple work process and works fine. Also you need SecWafLockOwner to set permission of the lock if you want to use in non-root process

### CODE REVIEW

File: standalone/server.c [hunk 1]
```diff
@@ -998,26 +998,14 @@ AP_DECLARE(void) unixd_pre_config(apr_pool_t *ptemp)
     apr_finfo_t wrapper;
 
 #if AP_SERVER_MAJORVERSION_NUMBER > 1 && AP_SERVER_MINORVERSION_NUMBER < 3
-#ifdef WAF_JSON_LOGGING_ENABLE
```
Comment (cp5555) [inline standalone/server.c:1001]: I think this code is not related to the persistent lock itself. It makes a little bit confusing.

File: apache2/Makefile.am [hunk 2]
```diff
@@ -6,6 +6,7 @@ mod_security2_la_SOURCES = acmp.c \
     ag_mdb/ag_mdb.cpp \
     waf_logging/waf_format.pb.cc \
     waf_logging/waf_log_util.cc \
+    waf_lock/waf_lock.cpp \
```
Comment (cp5555) [inline apache2/Makefile.am:9]: Can we separate the commit? One is for your standalone persistent lock utility code, another is for the ModSec-integration code (WAF JSON Log may be the independent one), and the other is for configuration and Makefile.

File: apache2/apache2_util.c [hunk 3]
```diff
@@ -289,24 +290,24 @@ static void get_ruleset_type_version(char* waf_ruleset_info, char* waf_ruleset_t
     }
 }
 
-static int write_file_with_lock(apr_global_mutex_t* lock, apr_file_t* fd, char* str) {
+static int write_file_with_lock(struct waf_lock* lock, apr_file_t* fd, char* str) {
```
Comment (cp5555) [inline apache2/apache2_util.c:293]: I don't know whether we need to keep the original code or not and whether we can directly change the API. Besides, If this code is only for the JSON log, could we separate it into one independent commit?

File: apache2/mod_security2.c [hunk 4]
```diff
@@ -833,7 +833,7 @@ static int hook_post_config(apr_pool_t *mp, apr_pool_t *mp_log, apr_pool_t *mp_t
  * Initialisation performed for every new child process.
  */
 static void hook_child_init(apr_pool_t *mp, server_rec *s) {
-    modsecurity_child_init(modsecurity);
+    modsecurity_child_init(mp, modsecurity);
```
Comment (cp5555) [inline apache2/mod_security2.c:836]: Do we need to change the ModSec API?

File: apache2/modsecurity.c [hunk 5]
```diff
@@ -154,85 +154,37 @@ int modsecurity_init(msc_engine *msce, apr_pool_t *mp) {
...
+    msce->wafjsonlog_lock = apr_pcalloc(mp, sizeof(struct waf_lock));
+    rc = Waf_createLock(msce->wafjsonlog_lock, "wafjsonlog_lock", strlen("wafjsonlog_lock"), msc_waf_lock_owner);    
+    if (Waf_lock_isError(rc)) {
         //ap_log_error(APLOG_MARK, APLOG_ERR, rv, s, "mod_security: Could not create modsec_wafjsonlog_lock");
         //return HTTP_INTERNAL_SERVER_ERROR;
         return -1;
     }
 #endif
 
-#if !defined(MSC_TEST)
```
Comment (cp5555) [inline apache2/modsecurity.c:174]: We remove too many code. For example, the micro MSC_TEST has been removed. I am not sure it is correct or not.

File: apache2/msc_logging.c [hunk 6]
```diff
@@ -1446,9 +1447,9 @@ void sec_audit_logger_json(modsec_rec *msr) {
     if (msr->txcfg->auditlog_type != AUDITLOG_CONCURRENT) {
 
         /* Unlock the mutex we used to serialise access to the audit log file. */
-        rc = apr_global_mutex_unlock(msr->modsecurity->auditlog_lock);
-        if (rc != APR_SUCCESS) {
-            msr_log(msr, 1, "Audit log: Failed to unlock global mutex: %s",
+	rc = Waf_freeExclusiveLock(msr->modsecurity->auditlog_lock);
```
Comment (cp5555) [inline apache2/msc_logging.c:1450]: alignment issue.

File: apache2/msc_logging.c [hunk 7]
```diff
@@ -1621,9 +1622,9 @@ void sec_audit_logger_native(modsec_rec *msr) {
 
     /* Lock the mutex, but only if we are using serial format. */
     if (msr->txcfg->auditlog_type != AUDITLOG_CONCURRENT) {
-        rc = apr_global_mutex_lock(msr->modsecurity->auditlog_lock);
-        if (rc != APR_SUCCESS) {
-            msr_log(msr, 1, "Audit log: Failed to lock global mutex: %s",
+	rc = Waf_getExclusiveLock(msr->modsecurity->auditlog_lock);
```
Comment (cp5555) [inline apache2/msc_logging.c:1625]: alignment issue.

File: apache2/msc_logging.c [hunk 8]
```diff
@@ -2226,9 +2227,9 @@ void sec_audit_logger_native(modsec_rec *msr) {
         sec_auditlog_write(msr, "\n", 1);
 
         /* Unlock the mutex we used to serialise access to the audit log file. */
-        rc = apr_global_mutex_unlock(msr->modsecurity->auditlog_lock);
-        if (rc != APR_SUCCESS) {
-            msr_log(msr, 1, "Audit log: Failed to unlock global mutex: %s",
+	rc = Waf_freeExclusiveLock(msr->modsecurity->auditlog_lock);
```
Comment (cp5555) [inline apache2/msc_logging.c:2230]: alignment issue.

File: apache2/waf_lock/murmur3.c [hunk 9]
```diff
@@ -0,0 +1,319 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cp5555) [inline apache2/waf_lock/murmur3.c:1]: Why do we need the murmur hash?

File: apache2/waf_lock/waf_lock_external.h [hunk 10]
```diff
@@ -0,0 +1,154 @@
+#ifndef _WAF_LOCK_EXTERNAL_HEADER
+#define _WAF_LOCK_EXTERNAL_HEADER
+
+#include "stdbool.h"
+#ifdef _WIN32
```
Comment (cp5555) [inline apache2/waf_lock/waf_lock_external.h:5]: Do we need to support Windows?

File: apache2/waf_lock/waf_lock_external.h [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+ ** return: WAF_LOCK_SUCCESS if successfully created or WAF_LOCK_ERROR if failed.
+ */
+int Waf_destroyLock(struct waf_lock *lock);
+
+/**
+ ** Close a lock.
+ ** @param lock: lock handler.
+ ** return: WAF_LOCK_SUCCESS if successfully created or WAF_LOCK_ERROR if failed.
+ */
+int Waf_cLoseLock(struct waf_lock *lock);
```
Comment (cp5555) [inline apache2/waf_lock/waf_lock_external.h:105]: Do we need to capitalize 'L' in 'Waf_cLoseLock'

File: apache2/waf_lock/waf_lock.cpp [hunk 12]
```diff
@@ -0,0 +1,533 @@
...
+    int read_lock_name_len = 0;
+    int write_lock_name_len = 0;
+    char* read_lock_name = NULL;
+    char* write_lock_name = NULL;
+    bool lock_exists = false;
+#endif
+    if (new_lock == NULL)
+        return WAF_LOCK_ERROR_HANDLE_NULL;
+#ifndef _WIN32
+    lock_id = Waf_lock_hash(lock_name, lock_name_length, DEFAULT_WAF_LOCK_ID_RANGE);
```
Comment (cp5555) [inline apache2/waf_lock/waf_lock.cpp:37]: I think we just need to do string and ID match. Do we need the murmur hash?

**Review by cp5555 — COMMENTED** [review]
I think we should separate the commits. One is for your standalone persistent lock utility code, another is for the ModSec-integration code (WAF JSON Log may be the independent one), and the other is for configuration and Makefile.

File: standalone/server.c (hunk 1 above)
Comment (azhao155) [reply standalone/server.c:1001]: This due to we need to set permission for the lock, since we have now own lock utility, so we don't need to have the old set permission code, as you say, i will split it into two pr, so it will be included in the integration of lock

File: apache2/Makefile.am (hunk 2 above)
Comment (azhao155) [reply apache2/Makefile.am:9]: yes

File: apache2/apache2_util.c (hunk 3 above)
Comment (azhao155) [reply apache2/apache2_util.c:293]: I think since it's a fix for the bug, we don't really need to keep the original code, will separate into another pr

File: apache2/mod_security2.c (hunk 4 above)
Comment (azhao155) [reply apache2/mod_security2.c:836]: since it's a fix and we do need to allocate for the struct, i think it's needed

File: apache2/modsecurity.c (hunk 5 above)
Comment (azhao155) [reply apache2/modsecurity.c:174]: good finding, i need to add MSC_TEST back, will fix it

File: apache2/msc_logging.c (hunk 6 above)
Comment (azhao155) [reply apache2/msc_logging.c:1450]: will do

File: apache2/msc_logging.c (hunk 7 above)
Comment (azhao155) [reply apache2/msc_logging.c:1625]: will do

File: apache2/msc_logging.c (hunk 8 above)
Comment (azhao155) [reply apache2/msc_logging.c:2230]: will do

File: apache2/waf_lock/waf_lock_external.h (hunk 10 above)
Comment (azhao155) [reply apache2/waf_lock/waf_lock_external.h:5]: yes

File: apache2/waf_lock/murmur3.c (hunk 9 above)
Comment (azhao155) [reply apache2/waf_lock/murmur3.c:1]: it seems to be an overkill, i will just let user input the sem id

File: apache2/waf_lock/waf_lock_external.h (hunk 11 above)
Comment (azhao155) [reply apache2/waf_lock/waf_lock_external.h:105]: will change

File: apache2/waf_lock/waf_lock.cpp (hunk 12 above)
Comment (azhao155) [reply apache2/waf_lock/waf_lock.cpp:37]: will let user input the id

File: apache2/Makefile.am (hunk 2 above)
Comment (asridharan) [reply apache2/Makefile.am:9]: +1 on breaking up the commits.

File: apache2/modsecurity.c [hunk 13]
```diff
@@ -154,85 +154,37 @@ int modsecurity_init(msc_engine *msce, apr_pool_t *mp) {
     curl_global_init(CURL_GLOBAL_ALL);
 #endif
     /* Serial audit log mutext */
-    rc = apr_global_mutex_create(&msce->auditlog_lock, NULL, APR_LOCK_DEFAULT, mp);
-    if (rc != APR_SUCCESS) {
+    msce->auditlog_lock = apr_pcalloc(mp, sizeof(struct waf_lock));
+    rc = Waf_createLock(msce->auditlog_lock, "auditlog_lock", strlen("auditlog_lock"), msc_waf_lock_owner);
```
Comment (asridharan) [inline apache2/modsecurity.c:158]: s/auditlog_lock/waf_auditlog_lock

File: apache2/modsecurity.c [hunk 14]
```diff
@@ -154,85 +154,37 @@ int modsecurity_init(msc_engine *msce, apr_pool_t *mp) {
...
         //return HTTP_INTERNAL_SERVER_ERROR;
         return -1;
     }
 
 #ifdef WAF_JSON_LOGGING_ENABLE
     /* Serial wafjson log mutext */
-    rc = apr_global_mutex_create(&msce->wafjsonlog_lock, NULL, APR_LOCK_DEFAULT, mp);
-    if (rc != APR_SUCCESS) {
+    msce->wafjsonlog_lock = apr_pcalloc(mp, sizeof(struct waf_lock));
+    rc = Waf_createLock(msce->wafjsonlog_lock, "wafjsonlog_lock", strlen("wafjsonlog_lock"), msc_waf_lock_owner);
```
Comment (asridharan) [inline apache2/modsecurity.c:168]: just for consistentcy can we do `waf_jsonlog_lock`? The prefix `waf_` implying the component its being used with.

File: apache2/modsecurity.c [hunk 15]
```diff
@@ -241,40 +193,32 @@ int modsecurity_init(msc_engine *msce, apr_pool_t *mp) {
...
 
-    if (msce->auditlog_lock != NULL) {
-        apr_status_t rc = apr_global_mutex_child_init(&msce->auditlog_lock, NULL, msce->mp);
-        if (rc != APR_SUCCESS) {
-            // ap_log_error(APLOG_MARK, APLOG_ERR, rs, s, "Failed to child-init auditlog mutex");
-        }
+    if (msce->auditlog_lock == NULL) {
+        msce->auditlog_lock = apr_pcalloc(mp, sizeof(struct waf_lock));
     }
+    Waf_createLock(msce->auditlog_lock, "auditlog_lock", strlen("auditlog_lock"), msc_waf_lock_owner);
```
Comment (asridharan) [inline apache2/modsecurity.c:203]: Is this idempotent? It just reuses the lock if it already exists?

File: apache2/waf_lock/murmur3.c (hunk 9 above)
Comment (asridharan) [reply apache2/waf_lock/murmur3.c:1]: +1

**Review by asridharan — CHANGES_REQUESTED** [review]

File: apache2/modsecurity.c (hunk 13 above)
Comment (azhao155) [reply apache2/modsecurity.c:158]: will do

File: apache2/modsecurity.c (hunk 14 above)
Comment (azhao155) [reply apache2/modsecurity.c:168]: ok

File: apache2/modsecurity.c (hunk 15 above)
Comment (azhao155) [reply apache2/modsecurity.c:203]: yes, if lock exists, it will reuse
